### PR TITLE
[4.0] Model classes improvement

### PR DESCRIPTION
**Pull request by joomdonation**
Pull Request for Issue # .

### Summary of Changes
This is redo of my original PR https://github.com/joomla/joomla-cms/pull/14567 because there was some changes to component dispatcher which affected the PR.

This PR contains improvement + clean up to model classes to support both namespace and none namespace components in Joomla 4. See https://github.com/joomdonation/joomla-cms/tree/namespace-com-content/administrator/components/com_content to see namespaced com_content component by @laoneo 

1. **Change to Model class**

- Change getName() method to support detect model name from both namespace class name and none namespace class name
- Change getTable() method to allow creating namespace table object
- Implement isCheckedOut to allow checking if a record is checked out (use in FormView class in my different PR)

2. **Change to Form Model class**
- Allow getting xml form and forms field from **Model** directory

3. **Change to List Model  class**
- Simplify the code of getFilterForm() method
- Allow getting xml form and forms field from **Model** directory

4. **Change to Admin Model class**
- Use getColumnAlias for check_out field instead of using hardcoded name (make it consistent with other places in our code base)

### Testing Instructions
This PR is for code review and discussion for Joomla 4 only. No human testing needed at the moment

### CODE REVIEW

File: libraries/cms/component/helper.php [hunk 1]
```diff
@@ -407,6 +407,42 @@ public static function renderComponent($option, $params = array())
...
+	 *
+	 * @param   string  $className  The class name of controller, model, view
+	 *
+	 * @return  string  The name of component
+	 *
+	 * @throws \Exception
+	 */
+	public static function getComponentName($className)
+	{
+		// In a namespace model class, the component name will be the part before Site/Administrator
```
Comment (yvesh) [inline libraries/cms/component/helper.php]: You can't assume that :-) Option should be always in the input.. or we need to have another way to inject it.

Comment (joomdonation) [inline libraries/cms/component/helper.php]: It is just a backup case only. At the moment, for controller, $option is getting from input directly. When create model , view classes, I also pass $option parameter to model and view constructor already

In our legacy MVC, we have code to detect component name fromclass name, too.

Comment (joomdonation) [inline libraries/cms/component/helper.php]: I wish we can force developers to pass option in the config array, that would make the code simpler and we don't have to detect from class name like this

**Comment by laoneo** [comment]
Should this pr not be against the 3.8 branch as we need that in the 3 series as well?

**Comment by yvesh** [comment]
@laoneo personally i feel we should go with 4 first and have everything working there. The dispatcher needs to be backported from 4 to 3.8, but as we are still working on it, is it worth the effort now?

**Comment by joomdonation** [comment]
Close because this PR conflict with the works by other dev. I will open smaller PRs contain improvements and not related to namespace MVC part.

**Pull request closed without merging**